File: bulk_renamer.js
// bulk_renamer.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const { promisify } = require('util');
const rename = promisify(fs.rename);
const stat = promisify(fs.stat);
const readdir = promisify(fs.readdir);

const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

async function getFiles(root, recursive, extensions, filterRegex) {
    const result = [];
    const entries = await readdir(root, { withFileTypes: true });
    for (const entry of entries) {
        const fullPath = path.join(root, entry.name);
        if (entry.isDirectory()) {
            if (recursive) {
                const sub = await getFiles(fullPath, recursive, extensions, filterRegex);
                result.push(...sub);
            }
            continue;
        }
        if (extensions && extensions.length) {
            const ext = path.extname(entry.name).slice(1).toLowerCase();
            if (!extensions.includes(ext)) continue;
        }
        if (filterRegex) {
            const re = new RegExp(filterRegex);
            if (!re.test(entry.name)) continue;
        }
        result.push(fullPath);
    }
    return result;
}

function applyTemplate(template, name, ext, num, prefix, suffix, replaceFrom, replaceTo, caseType) {
    let newName = name;
    if (replaceFrom !== undefined) {
        const re = new RegExp(replaceFrom, 'g');
        newName = newName.replace(re, replaceTo);
    }
    switch (caseType) {
        case 'lower': newName = newName.toLowerCase(); break;
        case 'upper': newName = newName.toUpperCase(); break;
        case 'title': newName = newName.replace(/\b\w/g, c => c.toUpperCase()); break;
    }
    if (prefix) newName = prefix + newName;
    if (suffix) newName = newName + suffix;

    // Replace variables
    const result = template.replace(/\{([^{}]+)\}/g, (match, key) => {
        if (key === 'name') return newName;
        if (key === 'ext') return ext;
        if (key === 'old') return name;
        if (key === 'date') return new Date().toISOString().slice(0,10);
        if (key.startsWith('num')) {
            if (key.includes(':')) {
                const fmt = key.split(':')[1];
                return require('util').format(`%${fmt}`, num);
            }
            return String(num);
        }
        return match;
    });
    return result;
}

async function main() {
    const args = require('minimist')(process.argv.slice(2), {
        string: ['t', 'p', 'e', 'f', 'prefix', 'suffix', 'replace-from', 'replace-to', 'case'],
        boolean: ['r', 'dry-run', 'v'],
        alias: { t: 'template', p: 'path', r: 'recursive', e: 'ext', f: 'filter' },
        default: { p: '.', 'dry-run': false, v: false }
    });

    const template = args.t || args.template;
    if (!template) {
        console.log(colorize('Ошибка: укажите шаблон (-t)', 'red'));
        process.exit(1);
    }

    const root = args.p || '.';
    const recursive = args.r || false;
    const extStr = args.e || '';
    const extensions = extStr ? extStr.split(',').map(s => s.trim()) : null;
    const filterRegex = args.f || '';
    const prefix = args.prefix || '';
    const suffix = args.suffix || '';
    const replaceFrom = args['replace-from'];
    const replaceTo = args['replace-to'];
    const caseType = args.case || '';
    const startNum = parseInt(args.number) || 1;
    const dryRun = args['dry-run'] || false;
    const verbose = args.v || false;

    let files;
    try {
        files = await getFiles(root, recursive, extensions, filterRegex);
    } catch (err) {
        console.log(colorize('Ошибка: ' + err.message, 'red'));
        process.exit(1);
    }

    if (files.length === 0) {
        console.log(colorize('Нет файлов.', 'yellow'));
        return;
    }

    let numCounter = startNum;
    let renamed = 0;

    for (const fullPath of files) {
        const dir = path.dirname(fullPath);
        const oldName = path.basename(fullPath);
        const ext = path.extname(oldName).slice(1);
        const nameWithoutExt = path.basename(oldName, '.' + ext);

        let newName = applyTemplate(template, nameWithoutExt, ext, numCounter, prefix, suffix, replaceFrom, replaceTo, caseType);
        numCounter++;

        if (!template.includes('{ext}') && ext) {
            newName = newName + '.' + ext;
        }

        const newPath = path.join(dir, newName);

        if (dryRun || verbose) {
            console.log(`${colorize('->', 'blue')} ${oldName} -> ${newName}`);
        }

        if (!dryRun) {
            // Conflict resolution
            try {
                await stat(newPath);
                // exists
                if (newPath !== fullPath) {
                    const base = path.basename(newName, path.extname(newName));
                    const ext2 = path.extname(newName);
                    let counter = 1;
                    let tryPath;
                    do {
                        tryPath = path.join(dir, `${base}_${counter}${ext2}`);
                        counter++;
                    } while (fs.existsSync(tryPath));
                    newPath = tryPath;
                    if (verbose) {
                        console.log(colorize(`  Конфликт, новое имя: ${path.basename(newPath)}`, 'yellow'));
                    }
                }
            } catch (err) {
                // file does not exist, ok
            }

            try {
                await rename(fullPath, newPath);
                renamed++;
            } catch (err) {
                console.log(colorize(`Ошибка при переименовании ${oldName}: ${err.message}`, 'red'));
            }
        }
    }

    if (dryRun) {
        console.log(colorize(`Симуляция завершена. Будет переименовано ${files.length} файлов.`, 'green'));
    } else {
        console.log(colorize(`Переименовано ${renamed} файлов.`, 'green'));
    }
}

main().catch(err => {
    console.log(colorize('Ошибка: ' + err.message, 'red'));
    process.exit(1);
});

File: bulk_renamer.rb
#!/usr/bin/env ruby
# bulk_renamer.rb
# encoding: UTF-8

require 'optparse'
require 'pathname'
require 'date'

COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

def get_files(root, recursive, extensions, filter_regex)
  files = []
  root = Pathname.new(root)
  return files unless root.exist?

  pattern = recursive ? '**/*' : '*'
  root.glob(pattern).each do |entry|
    next unless entry.file?
    name = entry.basename.to_s
    if extensions && !extensions.empty?
      ext = entry.extname[1..-1] || ''
      next unless extensions.include?(ext.downcase)
    end
    if filter_regex
      next unless name.match?(Regexp.new(filter_regex))
    end
    files << entry
  end
  files
end

def apply_template(template, name, ext, num, prefix, suffix, replace_from, replace_to, case_type)
  new_name = name.dup
  if replace_from && replace_to
    new_name.gsub!(Regexp.new(replace_from), replace_to)
  end
  case case_type
  when 'lower' then new_name.downcase!
  when 'upper' then new_name.upcase!
  when 'title' then new_name.gsub!(/\b\w/, &:capitalize)
  end
  new_name = prefix + new_name if prefix
  new_name = new_name + suffix if suffix

  result = template.dup
  result.gsub!(/\{([^{}]+)\}/) do |match|
    key = $1
    case key
    when 'name' then new_name
    when 'ext' then ext || ''
    when 'old' then name
    when 'date' then Date.today.strftime('%Y-%m-%d')
    when /^num/
      if key.include?(':')
        fmt = key.split(':')[1]
        sprintf("%#{fmt}", num)
      else
        num.to_s
      end
    else match
    end
  end
  result
end

options = {
  path: '.',
  recursive: false,
  extensions: nil,
  filter: nil,
  prefix: nil,
  suffix: nil,
  replace_from: nil,
  replace_to: nil,
  case_type: nil,
  start_num: 1,
  dry_run: false,
  verbose: false
}

parser = OptionParser.new do |opts|
  opts.banner = "Usage: bulk_renamer.rb -t <template> [options]"
  opts.on("-t", "--template TEMPLATE", "Шаблон") { |v| options[:template] = v }
  opts.on("-p", "--path PATH", "Путь") { |v| options[:path] = v }
  opts.on("-r", "--recursive", "Рекурсивно") { options[:recursive] = true }
  opts.on("-e", "--ext EXT1,EXT2", "Расширения") { |v| options[:extensions] = v.split(',').map(&:strip) }
  opts.on("-f", "--filter REGEX", "Фильтр") { |v| options[:filter] = v }
  opts.on("--prefix TEXT", "Префикс") { |v| options[:prefix] = v }
  opts.on("--suffix TEXT", "Суффикс") { |v| options[:suffix] = v }
  opts.on("--replace-from FROM", "Заменить от") { |v| options[:replace_from] = v }
  opts.on("--replace-to TO", "Заменить на") { |v| options[:replace_to] = v }
  opts.on("--case lower/upper/title", "Регистр") { |v| options[:case_type] = v }
  opts.on("--number N", Integer, "Начать нумерацию") { |v| options[:start_num] = v }
  opts.on("--dry-run", "Симуляция") { options[:dry_run] = true }
  opts.on("-v", "--verbose", "Подробно") { options[:verbose] = true }
  opts.on("-h", "--help", "Справка") { puts opts; exit }
end
parser.parse!

unless options[:template]
  puts colorize("Error: template required (-t)", :red)
  exit 1
end

files = get_files(options[:path], options[:recursive], options[:extensions], options[:filter])
if files.empty?
  puts colorize("No files found.", :yellow)
  exit 0
end

num_counter = options[:start_num]
renamed = 0

files.each do |full_path|
  old_name = full_path.basename.to_s
  ext = full_path.extname[1..-1] || ''
  name_without_ext = full_path.basename('.*').to_s

  new_name = apply_template(options[:template], name_without_ext, ext, num_counter,
                            options[:prefix], options[:suffix],
                            options[:replace_from], options[:replace_to],
                            options[:case_type])
  num_counter += 1

  unless options[:template].include?('{ext}')
    new_name = new_name + '.' + ext unless ext.empty?
  end

  new_path = full_path.dirname + new_name

  if options[:dry_run] || options[:verbose]
    puts "#{colorize('->', :blue)} #{old_name} -> #{new_name}"
  end

  unless options[:dry_run]
    # Conflict resolution
    if new_path.exist? && new_path != full_path
      base = new_path.basename('.*').to_s
      ext2 = new_path.extname
      counter = 1
      while true
        try_path = full_path.dirname + "#{base}_#{counter}#{ext2}"
        break unless try_path.exist?
        counter += 1
      end
      new_path = try_path
      if options[:verbose]
        puts colorize("  Conflict, new name: #{new_path.basename}", :yellow)
      end
    end

    begin
      File.rename(full_path, new_path)
      renamed += 1
    rescue => e
      puts colorize("Error renaming #{old_name}: #{e.message}", :red)
    end
  end
end

if options[:dry_run]
  puts colorize("Dry run completed. Would rename #{files.size} files.", :green)
else
  puts colorize("Renamed #{renamed} files.", :green)
end
